test/unit_test/runtime/magnitude_test.cpp

```diff
@@ -129,6 +135,69 @@ TEST_CASE("make_ratio performs prime factorization correctly")
   }
 }
 
+TEST_CASE("magnitude converts to numerical value")
+{
+  SECTION("Positive integer powers of integer bases give integer values")
+  {
+    constexpr auto mag_412 = as_magnitude<412>();
+    check_same_type_and_value(mag_412.value<int>, 412);
+    check_same_type_and_value(mag_412.value<std::size_t>, std::size_t{412});
+    check_same_type_and_value(mag_412.value<float>, 412.0f);
+    check_same_type_and_value(mag_412.value<double>, 412.0);
+  }
+
+  SECTION("Negative integer powers of integer bases compute correct values")
+  {
+    constexpr auto mag_0p125 = as_magnitude<ratio{1, 8}>();
+    check_same_type_and_value(mag_0p125.value<float>, 0.125f);
+    check_same_type_and_value(mag_0p125.value<double>, 0.125);
+  }
+
+  SECTION("pi to the 1 supplies correct values")
+  {
+    constexpr auto pi = pi_to_the<1>();
+    check_same_type_and_value(pi.value<float>, std::numbers::pi_v<float>);
+    check_same_type_and_value(pi.value<double>, std::numbers::pi_v<double>);
+    check_same_type_and_value(pi.value<long double>, std::numbers::pi_v<long double>);
+  }
+
+  SECTION("pi to arbitrary power performs computations in most accurate type at compile time")
+  {
+    constexpr auto pi_cubed = pi_to_the<3>();
+
+    auto cube = [](auto x) { return x * x * x; };
+    constexpr auto via_float = cube(std::numbers::pi_v<float>);
+    constexpr auto via_long_double = static_cast<float>(cube(std::numbers::pi_v<long double>));
+
+    constexpr auto pi_cubed_value = pi_cubed.value<float>;
+    REQUIRE(pi_cubed_value != via_float);
+    CHECK(pi_cubed_value == via_long_double);
+  }
+
+  SECTION("Impossible requests are prevented at compile time")
+  {
+    // Naturally, we cannot actually write a test to verify a compiler error.  But any of these can
+    // be uncommented if desired to verify that it breaks the build.
+
+    // (void)as_magnitude<412>().value<int8_t>;
```

Comment:
I use https://github.com/JohelEGP/jegp.cmake_modules/#jegpbuilderror to test that at build-time, if I can't use something like
```C++
export constexpr void expects(bool truth) noexcept {
  if (not truth) std::terminate();
}
export template<auto F> concept is_constant = requires { typename std::integral_constant<int, (F(), 0)>; };
static_assert(is_constant<[] {}>);
static_assert(is_constant<[] { return 0; }>);
static_assert(is_constant<[] { expects(true); }>);
static_assert(not is_constant<[] { expects(false); }>);
static_assert(
#if not defined(NDEBUG)
  not
#endif
  is_constant<[] { assert(false); }>);
static_assert(not is_constant<[] { throw 0; }>);
```

Reply:
I like this idea!  I was able to get these examples to work, but I couldn't figure out how to detect that, e.g., `as_magnitude<412>().value<int8_t>` shouldn't compile.  We could also try an external build system based approach, which is what I take the build rule in your first link to be (I didn't dive in in detail).

For now, I think it's fine to leave this as a follow up item.  The commented out cases aren't the best, but they do express intent in a way that's easy to check manually.